contracts/exchange-forwarder/contracts/src/MixinExchangeWrapper.sol

```diff
@@ -77,183 +80,243 @@ contract MixinExchangeWrapper is
+    /// @return Amounts filled and fees paid by maker and taker for this order.
+    function _marketBuySingleOrder(
+        LibOrder.Order memory order,
+        bytes memory signature,
+        uint256 makerAssetBuyAmount,
+        uint256 makerAssetAcquiredAmount
     )
         internal
-        returns (FillResults memory totalFillResults)
+        returns (LibFillResults.FillResults memory singleFillResults)
     {
-        bytes memory makerAssetData = orders[0].makerAssetData;
-        bytes memory wethAssetData = WETH_ASSET_DATA;
-
-        uint256 ordersLength = orders.length;
-        uint256 makerAssetFilledAmount = 0;
-        for (uint256 i = 0; i != ordersLength; i++) {
-
-            // We assume that asset being bought by taker is the same for each order.
-            // We assume that asset being sold by taker is WETH for each order.
-            orders[i].makerAssetData = makerAssetData;
-            orders[i].takerAssetData = wethAssetData;
-
-            // Calculate the remaining amount of makerAsset to buy
-            uint256 remainingMakerAssetFillAmount = _safeSub(makerAssetFillAmount, totalFillResults.makerAssetFilledAmount);
-
-            // Convert the remaining amount of makerAsset to buy into remaining amount
-            // of takerAsset to sell, assuming entire amount can be sold in the current order.
-            // We round up because the exchange rate computed by fillOrder rounds in favor
-            // of the Maker. In this case we want to overestimate the amount of takerAsset.
-            uint256 remainingTakerAssetFillAmount = _getPartialAmountCeil(
-                orders[i].takerAssetAmount,
-                orders[i].makerAssetAmount,
-                remainingMakerAssetFillAmount
+        // Percentage fee
+        if (order.takerFeeAssetData.equals(order.makerAssetData)) {
+            // Calculate the remaining amount of takerAsset to sell
+            uint256 remainingTakerAssetFillAmount = LibMath.getPartialAmountCeil(
+                order.takerAssetAmount,
+                order.makerAssetAmount.safeSub(order.takerFee),
+                makerAssetBuyAmount.safeSub(makerAssetAcquiredAmount)
             );
 
             // Attempt to sell the remaining amount of takerAsset
-            FillResults memory singleFillResults = _fillOrderNoThrow(
-                orders[i],
+            singleFillResults = _fillOrderNoThrow(
+                order,
                 remainingTakerAssetFillAmount,
-                signatures[i]
+                signature
+            );
+        // WETH fee
+        } else if (order.takerFeeAssetData.equals(order.takerAssetData)) {
+            // Calculate the remaining amount of takerAsset to sell
+            uint256 remainingTakerAssetFillAmount = LibMath.getPartialAmountCeil(
+                order.takerAssetAmount,
+                order.makerAssetAmount,
+                makerAssetBuyAmount.safeSub(makerAssetAcquiredAmount)
```

Comment:
This line is reused. Can we set it to `remainingMakerAssetFillAmount` above, similar to what we do in `_marketSellSingleOrder`?